save keras version & compile args when serializing models

cmyr opened the pull request:
I frequently load and retrain models serialized weeks and months earlier, and it's useful to know the `optimizer` and `loss` + other args that a model was compiled with. Other people have mentioned this as well.

Thought this was simple enough that I'd take a stab. This doesn't effect any existing functionality, but it lets me manually get the required information out of a json/yaml file.

I also took the liberty of saving the keras version number, which was mentioned in a `TODO` comment.

if @fchollet or anyone else has a better idea for how to do this (or why not to do this) I'm happy to revisit.

(fixes #2498)


edit: failing a pep8 test, I'll take a look in a bit and push again

fchollet commented on keras/engine/topology.py:
```diff
@@ -2348,6 +2346,41 @@ def load_weights(self, filepath):
             K.batch_set_value(weight_value_tuples)
         f.close()
 
+    def prepare_config(self):
```
I find that having `prepare_config` and `get_config` could be confusing... Maybe better naming would help?

fchollet commented on keras/engine/topology.py:
```diff
@@ -2348,6 +2346,41 @@ def load_weights(self, filepath):
+        }
+
+        if hasattr(self, 'optimizer'):
+            model_config['optimizer'] = prepare(self.optimizer)
```
Optimizers have a `get_config` method precisely for this purpose.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2348,6 +2346,41 @@ def load_weights(self, filepath):
+
+        if hasattr(self, 'optimizer'):
+            model_config['optimizer'] = prepare(self.optimizer)
+            model_config['loss'] = prepare(self.loss)
```
`__class__.__name__` should suffice here?

fchollet commented on keras/engine/topology.py:
```diff
@@ -2348,6 +2346,41 @@ def load_weights(self, filepath):
+        from six import string_types
+        from keras import __version__ as keras_version
+
+        def prepare(obj):
```
Not sure this is needed.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2367,11 +2400,7 @@ def get_json_type(obj):
-            'class_name': self.__class__.__name__,
-            'config': config,
-        }
+        model_config = self.prepare_config()
```
Sufficiently trimmed, `prepare_config` could fit in 5 lines, so maybe it's not so bad to have it twice? If factored out, it should have a better name anyway.

cmyr commented:
@fchollet alright, addressed some of your comments. Not sure if using the underbar for a name really fits code-style wise, but my imagination let me down. Happy to change if you have another idea.

fchollet commented:
> Not sure if using the underbar for a name really fits code-style wise, but my imagination let me down.

It's reasonable, it's used in many other parts of the code to indicate internal methods.

LGTM
